HttpServer/HttpServer/HttpServer.cs

```diff
@@ -0,0 +1,596 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net.Security;
+using System.Reflection;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.ResponseCompression;
+using Microsoft.AspNetCore.Server.Kestrel.Https;
+using Microsoft.Extensions.DependencyInjection;
+using Neo.IO.Json;
+
+
+namespace Neo.Plugins.HttpServer
+{
+    using Interceptor = Action<IHttpOperationPayload>;
+
+    public class HttpServer : Plugin, IHttpServer
```

Comment:
What is `IHttpServer`?

Reply:
Hi @erikzhang, the common code between RPC and Rest is the "HTTP engine". The IHttpServer is the "HTTP provider".
This implementation allows RPC and REST independently. You add interceptors to the requests and responses, allowing you to add RPC and REST endpoints independently.

Reply:
I need to see the interface details.

Reply:
I'm 90% confident (because I've seen the code) that this interface has only these 2 methods:
AddRequestInterceptor and AddResponseInterceptor.
Check https://github.com/neo-project/neo-plugins/pull/143/files#diff-506d69ca242c5fb9ebccbe04e0baa14f for current RPC implementation.

Reply:
This interface should be in the plugin, not in Neo.dll.

Reply:
If we do that, we are not going to allow different 'HTTP server' implementations. If that is ok for you, I agree.

Reply:
Actually, the interface _has_ to be on neo, otherwise, it won't be possible for people to implement it without adding a reference to this plugin.

Reply:
Plugins can also reference each other. I don't want all plugin interactions to depend on changes to Neo.

Reply:
Ok, I see the issue. I thought that we wanted to allow plugins to communicate without knowing that other plugins exist. If we expect plugins to reference other plugins, I agree that it can be moved to the plugins.

Reply:
But all plugins have an interface:
- ILogPlugin
- IMemoryPoolTxObserverPlugin
- IP2PPlugin
- IPersistencePlugin

So IHttpServer is just another plugin interface

Reply:
@erikzhang I am following the Plugin Implementation pattern, are you planning to change this pattern? I am pretty sure it would be impossible to have multiple implementation of HttpServer without this interface inside Neo.dll, it is just the way all plugins are working.

Reply:
Because the HttpServer is the base plugin, and Rpc and Rest are other plugins that depend on it.

Reply:
```C#
    public abstract class Plugin : IDisposable
    {
        public static readonly List<Plugin> Plugins = new List<Plugin>();
        private static readonly List<ILogPlugin> Loggers = new List<ILogPlugin>();
        internal static readonly List<IHttpServer> HttpServers = new List<IHttpServer>();
        protected static readonly List<IHttpPlugin> HttpPlugins = new List<IHttpPlugin>();
        internal static readonly List<IPersistencePlugin> PersistencePlugins = new List<IPersistencePlugin>();
        internal static readonly List<IP2PPlugin> P2PPlugins = new List<IP2PPlugin>();
        internal static readonly List<IMemoryPoolTxObserverPlugin> TxObserverPlugins = new List<IMemoryPoolTxObserverPlugin>();
```

```C#
protected Plugin()
        {
            Plugins.Add(this);

            if (this is ILogPlugin logger) Loggers.Add(logger);
            if (this is IP2PPlugin p2p) P2PPlugins.Add(p2p);
            if (this is IHttpServer rpcServer) HttpServers.Add(rpcServer);
            if (this is IHttpPlugin rpc) HttpPlugins.Add(rpc);
            if (this is IPersistencePlugin persistence) PersistencePlugins.Add(persistence);
            if (this is IMemoryPoolTxObserverPlugin txObserver) TxObserverPlugins.Add(txObserver);

            Configure();
        }
```

Reply:
`HttpPlugin` can reference to `HttpServer` directly. So it doesn't need an interface. In this way we don't need to modify Neo.

Reply:
@erikzhang I could drop IHttpServer but then the Http Plugins could only work with this HttpServer implementation, not a generic one, I can easily do that if that is your intention.

Reply:
I really don't think we need multiple HttpServer implementations.

Reply:
Alright, so I will change that, thanks for the clarification.

Reply:
Done